### DISCUSSION TITLE
Examples of Python API usage

### DISCUSSION THREAD

**Discussion by kaon1** [cea-sec/ivre#1463] [Q&A] [answered]
Hi, I am looking to get started with IVRE and I imported some test data. 

I would like to query the data via the python API but I cannot find example code to get started. I'm looking for simple queries like

- List all open ports
- List all ip addresses with port 443 open 
- etc

Is there any documentation on how to get started with the python api for ivre?

Thank you

**Comment by p-l-** [comment, marked as answer]
Hi there,

The Python is mostly documented using types and dosctrings for now. If you want to query the view database, the object `ivre.db.db.view` will be used, and you can run, for example:
```python
from ivre.db import db
help(db.view)
```

If you want to list all the open ports, you can run:
```python
from ivre.db import db
db.view.distinct("ports.port")  # this is a generator, you can iterate over it
list(db.view.distinct("ports.port"))  # this will build a list

[80, 23, 443]
```

You can also use `.topvalues()` (so that you have both protocol and port number):
```python
from ivre.db import db
db.view.topvalues("port:open", limit=None)  # this is a generator, you can iterate over it
list(db.view.topvalues("port:open", limit=None))  # this will build a list

[{'count': 809, '_id': ('tcp', 443)}, {'count': 219, '_id': ('tcp', 80)}, {'count': 99, '_id': ('tcp', 23)}
```

For the IP addresses with port (TCP/)443 open:
```python
db.view.distinct("addr", flt=db.view.searchport(443))  # this is a generator, you can iterate over it
list(db.view.distinct("addr", flt=db.view.searchport(443)))  # this will build a list

[::1", "127.0.0.1"]
```

**Comment by kaon1** [comment]
Hi @p-l- after doing some testing I notice the diff_categories function is not returning the ipv4 address. I see two 'addr' fields that start with -9 but I can't translate these values to my IPv4 address.

Here's an example:

`list(db.view.diff_categories(category1="scan_1",category2="scan_2 ",include_both_open=False))`

result example:
`{'addr': [-9223...<redacted>, -9223...<redacted>], 'proto': 'tcp', 'port': 80, 'value': -1}`

When I try to decode these -9 numbers I get an error:

```
>>> ivre.utils.int2ip(-9223...<redacted>)
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/ivre/utils.py", line 255, in int2ip
    return socket.inet_ntoa(struct.pack("!I", ipint))
struct.error: argument out of range
```

Am I missing something? Should I be able to convert this -9 addr to an IP?

**Reply by p-l-** [reply]
That's probably a bug in .diff_categories(). You may want to open an issue, using example IP addresses so that you can share the values (that will make it easier to find and fix the bug).

**Reply by p-l-** [reply]
Nevermind. I have pushed a PR to fix that bug; see #1467.
